src/compiler/checker.ts

```diff
@@ -27869,7 +27898,23 @@ namespace ts {
                 }
                 name = idText(name);
             }
-            return getSpellingSuggestionForName(name, props, SymbolFlags.Value);
+            const suggestion = getSpellingSuggestionForName(name, props, SymbolFlags.Value);
+            if (suggestion) {
+                return suggestion;
+            }
+            // If we have `#typo in expr` then we can still look up potential privateIdentifiers from the surrounding classes
+            if (typeof originalName !== "string" && isPrivateIdentifierInInExpression(originalName.parent)) {

```

Comment:
This only applies to private identifiers, right? In that case wouldn't it be more precise to check `isPrivateIdentifier(originalName)` instead of `typeof originalName !== "string"`?

On the other hand, isn't the comment true of all properties, private or not? In that case, the whole new block could apply both to private property declarations and normal property declarations.

Reply:
Actually suggestions already work for nested property accesses in this program:

```ts
class C {
    #patisserie = 1
    m() {
        const nested = class {
            #boulangerie = 2
            m2(c: C) {
                this.#oulangerie
                c.#atisserie
            }
        }
    } 
}
```

So I don't understand why special handling is needed for PrivateIdentifierInInExpressions.

Reply:
The special handling is because property access appears to only look up the suggestions based on the receiver. 

```ts
class C {
    #patisserie = 1
    m() {
        const nested = class {
            #boulangerie = 2
            m2(this: any, c: any) {
                this.#oulangerie // no spelling suggestions here
                c.#atisserie // or here
            }
        }
    } 
}
```

The use case of `PrivateIdentifierInInExpressions` is specifically for when the receiver's type is wide, potentially `unknown`. So instead we look up the suggestions lexically.

Happy to discuss further if we think this should be expanded to unsafe (untyped) private-field property access too.

Reply:
Thanks, that example is a good explanation.
I intentionally limited property access lookup by the type of the receiver in order to keep precision high. So the example is working as intended. Let's move lexical lookup to a separate issue+PR pair. I think it might make sense as a separate feature [1], but not as part of this one.

[1] Some questions I'd like to see discussed there:

1. How much does it improve real-world recall?
2. How much does it hurt precision?
3. Should it only apply to #privates?
4. Does iterated lookup hurt performance? (Probably not, but worth thinking about a little.)
5. When the type is explicitly any or unknown, shouldn't the editor first suggest a specific type or narrowing for the receiver?

Reply:
code-fix removed from this PR and separate issue raised: https://github.com/microsoft/TypeScript/issues/45908